source/WeightedList.h

```diff
@@ -69,20 +75,56 @@ class WeightedList {
 
 private:
 	std::vector<Type> choices;
+	std::vector<std::size_t> weights;
 	std::size_t total = 0;
 };
 
 
 
+template <class T>
+std::size_t erase(WeightedList<T> &list, const T &item)
+{
+	return erase_if(list, [&item](const T &t) noexcept -> bool { return item == t; });
+}
+
+
+
+template <class T, class UnaryPredicate>
+std::size_t erase_if(WeightedList<T> &list, UnaryPredicate pred)
+{
+	std::size_t erased = 0;
+	auto it = list.choices.begin();
+	auto wit = list.weights.begin();
+	while(it != list.choices.end())
+	{
+		if(pred(*it))
+		{
+			list.total -= *wit;
+			wit = list.weights.erase(wit);
+			it = list.choices.erase(it);
```

Comment:
Repeatedly invoking `vector::erase` is suboptimal; this invokes the move or copy assignment operator on each following element each time the predicate is true.

We should instead only invoke the assignment operator at most once for each element to be retained.

Reply:
So should we `end = remove(it, end, *it)` (and the same for wit) for each match and then erase at the end?

Reply:
For the general container case, you would want to store a list of bad indices as you scan forwards, removing from this list as you find "good" indices whose contents should be preserved. After moving the content from "first good" to "first bad," that moved-from index is now considered "bad."

However, our specific container has no need to preserve element order, so we only need to move elements when there are no remaining elements to move into a bad index ahead of them. To minimize work, we can fill from the end inward.


Ex:
A wl has 5 elements. If the first and fourth are "bad," we can just move-assign the last element to the first position, rather than shift 2 and 3 down by one and 5 down by 2.

Reply:
I'm not sure if this was what tehhowch is referring to, but the easiest way to implement this is by doing two remove_ifs at the same time and then erasing the removed elements at the end.

Reply:
<details><summary>Test program.</summary>

```
#include <iostream>
#include <iterator>
#include <numeric>
#include <vector>

int main ()
{
	std::vector<int> choices = {1,2,3,4,5,6,7,8,9,10};
	std::vector<int> weights = {1,2,3,4,5,6,7,8,9,10};
	int total = std::accumulate(weights.begin(), weights.end(), 0);
	std::cout << "The beginning total is: " << total << '\n';

	auto pred = [](const int &i) noexcept -> bool { return i % 2; }; 
	std::size_t erased = 0;
	unsigned available = choices.size() - 1;
	for(unsigned index = 0; index < choices.size() - erased; )
		if(pred(choices[index]))
		{
			total -= weights[index];
			std::cout << "Replacing " << choices[index] << " with " << choices[available] << '\n';
			choices[index] = std::move(choices[available]);
			weights[index] = std::move(weights[available]);
			--available;
			++erased;
		}
		else
			++index;

	++available
	choices.erase(std::next(choices.begin(), available), choices.end());
	weights.erase(std::next(weights.begin(), available), weights.end());

	std::cout << "Number erased: " << erased << '\n';
	std::cout << "Final total: " << total << '\n';
	std::cout << "The remaining choices are:";
	for (auto &x: choices)
		std::cout << ' ' << x;
	std::cout << '\n';

	return 0;
}
```

</details>

<details><summary>Test output.</summary>

```
The beginning total is: 55
Replacing 1 with 10
Replacing 3 with 9
Replacing 9 with 8
Replacing 5 with 7
Replacing 7 with 6
Number erased: 5
Final total: 30
The remaining choices are: 10 2 8 4 6
```

</details>

Reply:
Make sure `choices[available]` is to be kept before you use it.

Writing the unit tests first will help you refactor.

Reply:
Take two:
<details><summary>Test program.</summary>

```
#include <iostream>
#include <iterator>
#include <numeric>
#include <utility>
#include <vector>

int main ()
{
	std::vector<int> choices = {1,2,3,4,5,6,7,8,9,10,11,12};
	std::vector<int> weights = {1,2,3,4,5,6,7,8,9,10,11,12};
	int total = std::accumulate(weights.begin(), weights.end(), 0);
	std::cout << "The beginning total is: " << total << '\n';

	auto pred = [](const int &i) noexcept -> bool { return i % 2; }; 
	std::size_t erased = 0;
	unsigned index = 0;
	unsigned available = choices.size() - 1;
	while(index < choices.size() - erased)
		if(pred(choices[index]))
		{
			std::cout << choices[index] << " matches pred\n";
			total -= weights[index];
			if(index != available)
			{
    			while(pred(choices[available]))
    			{
    			    std::cout << choices[available] << " matches pred\n";
    			    total -= weights[available];
    	            --available;
    	            ++erased;
    			}
    			std::cout << "Replacing " << choices[index] << " with " << choices[available] << '\n';
    			choices[index] = std::move(choices[available]);
    			weights[index] = std::move(weights[available]);
    		    --available;
            }
			++erased;
		}
		else
			++index;

	choices.erase(std::prev(choices.end(), erased), choices.end());
	weights.erase(std::prev(weights.end(), erased), weights.end());

	std::cout << "Number erased: " << erased << '\n';
	std::cout << "Final total: " << total << '\n';
	std::cout << "The remaining choices are:";
	for (auto &x: choices)
		std::cout << ' ' << x;
	std::cout << '\n';
	std::cout << "The remaining weights are:";
	for (auto &x: weights)
		std::cout << ' ' << x;
	std::cout << '\n';

	return 0;
}
```

<details>

<details><summary>Test output.</summary>

```
The beginning total is: 78
1 matches pred
Replacing 1 with 12
3 matches pred
11 matches pred
Replacing 3 with 10
5 matches pred
9 matches pred
Replacing 5 with 8
7 matches pred
Number erased: 6
Final total: 42
The remaining choices are: 12 2 10 4 8 6
The remaining weights are: 12 2 10 4 8 6
```

</details>